STITCH-2211 Swift SDK: Create data models and implement local event coalescence

jsflax opened the pull request:
# "Port" the following classes:
## InstanceSynchronization:
- InstanceSynchronization now acts as a container for an internal config (`InstanceSynchronization.Config`). This was due to the fact that the current Java equivalent can exist in a partial state, which is a no-go in Swift. It also cleans up a significant amount of the encoding/decoding code, which was cumbersome. IMO, we should make these changes on Java as well.
- Conformance to `Sequence`. Can iterate through `NamespaceSynchronization.Configs`.
- Subscript access. `InstanceSynchronization` subscript uses a mutating getter to either get existing `NamespaceSynchronization`s or will create new ones.
## NamespaceSynchronization
- See above for container information.
- Conformance to `Sequence`. Can iterate through `CoreDocumentSynchronization.Configs`.
- Subscript access. `NamespaceSynchronization` subscript will either fetch existing `CoreDocumentSynchronization` by `id` or can set new ones by id.
## CoreDocumentSynchronization
- See above for container information. `Config` properties are now private. The properties are exposed in the outlying struct `CoreDocumentSynchronization` as computed properties. These getters/setters are synchronized via a `ReadWriteLock`, with the setters mutating the underlying `config`.
## ChangeEvent
- `OperationType` has been moved out to its own enum. 
    - The string values of each case are now implicit.
- `transform` (the method that converts a ChangeEvent with an associated type of `Document` to a user associated type) now takes on a more Swifty shape.
## UpdateDescription
- `UpdateDescription` has been moved out to its own class.
- Conversion to a `Document` is now computed lazily.
- `diff` is now a method extended to the `Document` class.
## ConflictHandler
- Uses associated types, which forces us to store locally using type erasure.
## ChangeEventListener
- Uses associated types, which forces us to store locally using type erasure.
## MongoNamespace
- Not sure why this class doesn't exist on `MongoMobile`. Simple port-over.
## DocumentVersionInfo
- `nextVersionDocument` is now lazily computed.
## ReadWriteLock
- Swift doesn't have a default RWLock, so this is a simple wrapper above `pthread_rwlock_t`.

## Other additions, notable or otherwise:
- More comments.
- More tests.
- Take advantage of nullability and mutability.
- More Swift.

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+     Generates a change event for a local update of a document in the given namespace referring
+     to the given document _id.
+
+     - parameter namespace: the namespace where the document was inserted.
```
[nit] updated

dkaminsky replied:
+1, should be [typo] and should be fixed

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+     Generates a change event for a local replacement of a document in the given namespace referring
+     to the given document _id.
+
+     - parameter namespace: the namespace where the document was inserted.
```
[nit] replaced

dkaminsky replied:
+1, should be [typo] and should be fixed

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+     Generates a change event for a local deletion of a document in the given namespace referring
+     to the given document _id.
+
+     - parameter namespace: the namespace where the document was inserted.
```
[nit] deleted

dkaminsky replied:
+1, should be [typo] and should be fixed

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+     to the given document _id.
+
+     - parameter namespace: the namespace where the document was inserted.
+     - parameter documentId: the _id of the document that was updated.
```
[nit] deleted

dkaminsky replied:
+1, should be [typo] and should be fixed

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+     to the given document _id.
+
+     - parameter namespace: the namespace where the document was inserted.
+     - parameter documentId: the _id of the document that was updated.
```
[nit] replaced

dkaminsky replied:
+1, should be [typo] and should be fixed

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/NamespaceSynchronizationConfig.swift:
```diff
@@ -0,0 +1,184 @@
+     - parameter conflictHandler: a ConflictHandler to handle conflicts on this namespace
+     - parameter changeEventListener: a ChangeEventListener to listen to events on this namespace
+     */
+    mutating func configure<T: ConflictHandler, V: ChangeEventListener>(conflictHandler: T,
```
[api discussion] In Java, this pattern works nicely, because you can pass an implementation of an interface inline (or use a lambda). In Swift, you can't really do that AFAIK. I know this isn't where we're going to expose configuration for the user, but have we thought about having configure variants that accept closures instead of protocols?

jsflax replied:
I think it's a good idea to do both. We can store it either way as their type erased equivalents, and the user can go about however they place. IMO, in an actual application, there's a stronger use case for having a singleton conform to these protocols than just a throwaway lambda. However, for the sake of the user, probably best to support both.

jsflax replied:
This will be exposed at a higher level.

jsflax commented on Core/Services/StitchCoreAWSS3Service/StitchCoreAWSS3Service.xcodeproj/project.pbxproj:
```diff
@@ -10,8 +10,8 @@
 		492FA73321909351001DFD53 /* StitchCoreSDKMocks.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 492FA73221909351001DFD53 /* StitchCoreSDKMocks.framework */; };
-		AE3FAAFB31BE9D7D55FFC622 /* libPods-StitchCoreAWSS3Service.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 471CBEEEDD51986508C0F591 /* libPods-StitchCoreAWSS3Service.a */; };
-		FDA5BEF356900B013C812B0F /* libPods-StitchCoreAWSS3ServiceTests.a in Frameworks */ = {isa = PBXBuildFile; fileRef = 85603B760FB854A36EB63122 /* libPods-StitchCoreAWSS3ServiceTests.a */; };
+		5CADF451D4021A34E0C51407 /* Pods_StitchCoreAWSS3Service.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = FCBD14893738BF6757F05B85 /* Pods_StitchCoreAWSS3Service.framework */; };
```
Because we moved to frameworks with our podfile. Generally, any update to the podfile will cause pbxproj files to change.

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/BSONValue+Hashable.swift:
```diff
@@ -0,0 +1,81 @@
+    case (let value as String):
+        hasher.combine(value)
+    case (let value as RegularExpression):
+        hasher.combine(value.options)
```
I wrote this based on how the driver wrote their equality function, but I can change it. Hasher just accumulates the hashes, it's nothing special.

adamchel replied:
ah kk, no need to change then, esp. if this code is going away to do whatever the driver does

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+    /// The namespace (database and or collection) affected by the event.
+    public let ns: MongoNamespace
+    /**
+     The ObjectID of the document created or modified by the
```
I lifted this right from the docs. https://docs.mongodb.com/manual/reference/change-events/

dkaminsky replied:
+1 adam

Prefer the more generic `_id` or, if you'd rather not use field name, "unique BSON identifier" or something along thos elines

adamchel replied:
ah in that case the docs are wrong, I'll message Caleb

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+    /// Whether or not this ChangeEvent has pending writes.
+    public let hasUncommittedWrites: Bool
+
+    public static func == (lhs: ChangeEvent<DocumentT>, rhs: ChangeEvent<DocumentT>) -> Bool {
```
I don't know why that would be the case. DocumentKey should be unique, and we shouldn't be comparing ChangeEvents between namespaces.

jsflax replied:
Spoken about offline.

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+    static func transform<T: Codable>(changeEvent: ChangeEvent<Document>) -> ChangeEvent<T> {
+        var fullDocument: T?
+        if let actualFullDocument = changeEvent.fullDocument {
+            fullDocument = try! BSONDecoder().decode(T.self,
```
More of a [style] than a [nit].

adamchel commented on publish_pods.sh:
```diff
@@ -0,0 +1,21 @@
+pod trunk push Core/StitchCoreSDK/StitchCoreSDK.podspec --allow-warnings --verbose
```
[minor] if my docs stuff goes in before this, add `contrib/publish_docs.sh` before pushing any of the pods. that script will generate and publish the docs as long as you have the right AWS keys set up in your env

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/ChangeEventUnitTests.swift:
```diff
@@ -0,0 +1,112 @@
+        let actualLocalInsertEvent = ChangeEvent<Document>.changeEventForLocalInsert(namespace: namespace,
+                                                                                     document: document,
+                                                                                     writePending: writePending)
+        XCTAssertEqual(expectedLocalInsertEvent, actualLocalInsertEvent)
```
since this uses the the incomplete `==` implementation for change events, this actually isn't comparing all the fields of the expected and actual change event

dkaminsky replied:
I think this is resolved with the fixed == implementation

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/ChangeEventUnitTests.swift:
```diff
@@ -0,0 +1,112 @@
+
+        let actualLocalReplaceEvent = ChangeEvent<Document>.changeEventForLocalReplace(namespace: namespace, documentId: docId, document: document, writePending: writePending)
+
+        XCTAssertEqual(expectedLocalReplaceEvent, actualLocalReplaceEvent)
```
[nit] ditto

dkaminsky replied:
ditto

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/ChangeEventUnitTests.swift:
```diff
@@ -0,0 +1,112 @@
+            fullDocumentAfterUpdate: document,
+            writePending: writePending)
+
+        XCTAssertEqual(expectedLocalUpdateEvent, actualLocalUpdateEvent)
```
[nit] ditto

dkaminsky replied:
ditto

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/ChangeEventUnitTests.swift:
```diff
@@ -0,0 +1,112 @@
+            updateDescription: nil,
+            hasUncommittedWrites: writePending)
+
+        let actualLocalDeleteEvent = ChangeEvent<Document>.changeEventForLocalDelete(namespace: namespace,
```
[nit] ditto

dkaminsky replied:
ditto

dkaminsky commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/CoreDocumentSynchronizationConfig.swift:
```diff
@@ -0,0 +1,338 @@
+            isStale = true
+        }
+
+        docLock.writeLock()
```
[major] shouldn't we acquire write lock before changing paused/stale status?

jsflax replied:
I ported over what happens on Android and assumed we had a reason for doing it this way. @edaniels?

jsflax replied:
This is handled implicitly in the setters. I can make it explicit, however this is not a design flaw.

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/DocumentVersionInfoUnitTests.swift:
```diff
@@ -0,0 +1,36 @@
+@testable import StitchCoreRemoteMongoDBService
+
+class DocumentVersionInfoUnitTests: XCTestCase {
+    func testVersionDocument() {
```
[minor] also add `testGetNextVersion` from the Android unit tests

jsflax replied:
I am technically testing the next version here.

dkaminsky commented on Podfile:
```diff
@@ -1,19 +1,22 @@
 workspace 'Stitch.xcworkspace'
```
[q] Should this be committed and merged to master prior to the rest of this review, bypassing sync branch?

jsflax replied:
We had to change dependencies for this PR.

dkaminsky commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DocumentVersionInfo.swift:
```diff
@@ -0,0 +1,159 @@
+import MongoSwift
```
[nit] this wasn't in scope for this ticket IIRC

dkaminsky commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/DocumentVersionInfo.swift:
```diff
@@ -0,0 +1,159 @@
+    lazy var nextVersion: Document = {
+        guard hasVersion,
+            let versionDoc = versionDoc,
+            let version = version else {
```
[minor] isn't `hasVersion` redundant with `let version = version` based on the definition of hasVersion below?

jsflax replied:
Good point, but that raises other questions of how this is setup.

dkaminsky replied:
which are..? :)

jsflax replied:
Questions that I stated we should kick downfield until the versioning ticket. Different semantics means there's a conversation worth being had about the original implementation.

dkaminsky commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/NamespaceSynchronizationConfig.swift:
```diff
@@ -0,0 +1,184 @@
+
+    /// Allows for the iteration of the document configs contained in this instance.
+    struct NamespaceSynchronizationIterator: IteratorProtocol {
+        typealias Element = CoreDocumentSynchronization
```
[q] not redundant with wider-scoped typealias above?

adamchel replied:
when you implement a Swift protocol with associated types (the half-baked Swift feature that lets you do generic interfaces), it's generally idiomatic to use typealias to specify the type parameter of the protocol you're implementing

dkaminsky replied:
kk, so even though it's the same typealias as above, this has some semantic meaning? or is it purely a style thing?

dkaminsky commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/CoreRemoteMongoClientUnitTests.swift:
```diff
@@ -1,8 +1,10 @@
 @testable import StitchCoreRemoteMongoDBService
 
-final class CoreRemoteMongoClientUnitTests: XCTestCase {
+final class CoreRemoteMongoClientUnitTests: XCMongoMobileTestCase {
```
[q] this seems more a mongo mobile change not a sync hange?

jsflax replied:
We were doing this improperly before. Unfortunately, MongoMobile is setup in a way that it can only be initialized once and torndown once. So we need to be very careful about where we do those operations, especially with test bundles.

dkaminsky commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/ChangeEventUnitTests.swift:
```diff
@@ -0,0 +1,112 @@
+
+        XCTAssertEqual(expectedLocalDeleteEvent, actualLocalDeleteEvent)
+    }
+}
```
[major] We should pin the error behavior for unknown with a test to make sure that all exception states are deliberate rather than implicit

dkaminsky commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/XCMongoMobileTestCase.swift:
```diff
@@ -0,0 +1,41 @@
+import Foundation
```
[q] This seems more a mongo mobile thing than a sync thing?

jsflax replied:
See comment above.

jsflax commented:
Disregard that comment

On Thu 15 Nov 2018, 12:58 Jason Flax, <notifications@github.com> wrote:

> *@jsflax* commented on this pull request.
> ------------------------------
>
> In
> Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/CoreDocumentSynchronizationConfig.swift
> <https://github.com/mongodb/stitch-ios-sdk/pull/103#discussion_r233948518>
> :
>
> > +     locally emitted change event. This variant maintains the last version set.
> +
> +     - parameter atTime: the time at which the write occurred.
> +     - parameter changeEvent: the description of the write/change.
> +     */
> +    mutating func setSomePendingWrites(atTime: TimeInterval,
> +                                       changeEvent: ChangeEvent<Document>) {
> +        // if we were frozen
> +        if (isPaused) {
> +            // unfreeze the document due to the local write
> +            isPaused = false
> +            // and now the unfrozen document is now stale
> +            isStale = true
> +        }
> +
> +        docLock.writeLock()
>
> I ported over what happens on Android and assumed we had a reason for
> doing it this way. @edaniels <https://github.com/edaniels>?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mongodb/stitch-ios-sdk/pull/103#discussion_r233948518>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEJ7QEaXQkjHDJWTw8kICeszA-LbrNK9ks5uvasogaJpZM4YcyA1>
> .
>

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/StitchCoreRemoteMongoDBService.xcodeproj/StitchCoreRemoteMongoDBServiceTests_Info.plist:
```diff
@@ -1,25 +1,26 @@
+	<key>CFBundleVersion</key>
+	<string>$(CURRENT_PROJECT_VERSION)</string>
+	<key>NSPrincipalClass</key>
+	<string>$(PRODUCT_NAME).XCMongoMobileConfiguration</string>
```
This class is ran before and after test bundles are run. This is required to allow MongoMobile to start and teardown cleanly.

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/CoreDocumentSynchronizationConfig.swift:
```diff
@@ -0,0 +1,338 @@
+}
+
+/**
+ The synchronization portal for this document.
```
What would you say is a better choice of word? I had some trouble choosing.

adamchel replied:
Maybe something along the lines of a "a class containing information about the synchronization state of a particular document."

dkaminsky replied:
+1 portal is a bad name -- is it a handle? a state manager? a state API?

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/CoreDocumentSynchronizationConfig.swift:
```diff
@@ -0,0 +1,338 @@
+    /// The configuration for this document.
+    private(set) var config: Config
+    /// The namespace this document is stored in.
+    var namespace: MongoNamespace { get { return config.namespace } }
```
More of an explicit style choice

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/CoreDocumentSynchronizationConfig.swift:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    init(docsColl: MongoCollection<CoreDocumentSynchronization.Config>,
+         config: inout Config) {
```
The configs should stay mutable so that the in memory mapping match what is in the local store.

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/MongoNamespace.swift:
```diff
@@ -0,0 +1,32 @@
+
+    public init(from decoder: Decoder) {
+        let description = try! decoder.singleValueContainer().decode(String.self)
+        let substrings = description.split(separator: ".")
```
We'll have to make it a nullable initializer then.

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/NamespaceSynchronizationConfig.swift:
```diff
@@ -0,0 +1,184 @@
+            self.syncedDocuments = syncedDocuments
+        }
+        static func == (lhs: NamespaceSynchronization.Config, rhs: NamespaceSynchronization.Config) -> Bool {
+            return true
```
Oops

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ReadWriteLock.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @discardableResult
+    func unlock() -> Bool {
```
This was the question I brought up in Slack the other night when writing the RW lock. The unlock statement is smart enough to know the state of itself w.r.t read or write.

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -0,0 +1,168 @@
+
+/// The type of operation that occurred within a ChangeEvent.
+public enum OperationType: String, Codable {
+    case insert, delete, replace, update, unknown
```
Invalidate?

adamchel replied:
ah discussed with doug offline. `invalidate` is an event type that exists for MongoDB change events, but it doesn't exist for us since we close the client stream when we get an invalidate event on the Stitch server, so we don't need it. My other comment about the unknown event still stands though

jsflax commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEventListener.swift:
```diff
@@ -0,0 +1,31 @@
+/**
+ ChangeEventListener receives change event notifications.
+ */
+public protocol ChangeEventListener {
```
That would be a choice we could make, but then we would remove the ability for users to configure using lambdas. They can do a switch statement over the operation type enum to replicate that same behavior.

jsflax commented:
@adamchel @dkaminsky I think we should kick the versioning changes downfield until we get to that ticket. I've laid down the base model which arguably was in scope for this one.

Otherwise, I've addressed your remarks. PTAL.

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -2,13 +2,22 @@ import MongoSwift
 /// The type of operation that occurred within a ChangeEvent.
 public enum OperationType: String, Codable {
-    case insert, delete, replace, update, unknown
+    case insert = "INSERT"
```
[minor] I think they should actually be lowercase

jsflax replied:
Are we sure?

adamchel replied:
https://github.com/mongodb/stitch-android-sdk/blob/master/core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/ChangeEvent.java#L110

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -2,13 +2,22 @@ import MongoSwift
+
     /// Metadata related to the operation.
-    public let id: Document
+    public let id: AnyBSONValue
```
[q] why did this change? this is going to always be a document

dkaminsky replied:
Is it? You can't have a BSON String or Integer ID?

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ChangeEvent.swift:
```diff
@@ -44,12 +53,54 @@ public struct ChangeEvent<DocumentT: Codable>: Codable, Hashable {
+        self.id = try container.decode(AnyBSONValue.self, forKey: .id)
+        self.operationType = OperationType.init(
+            rawValue: try container.decode(String.self, forKey: .operationType)
+        ) ?? .unknown
```
[minor] should be `try?`

jsflax replied:
If the operationType key should always be there, this should fail if the key doesn't exist.

adamchel replied:
Oh I see, this would only fail if `container.decode` fail if it's not a string. In that case that's fine

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/ConflictHandler.swift:
```diff
@@ -19,25 +19,27 @@ public protocol ConflictHandler {
+                    documentId: documentId,
+                    localEvent: try ChangeEvent<U.DocumentT>.transform(changeEvent: localEvent),
+                    remoteEvent: try ChangeEvent<U.DocumentT>.transform(changeEvent: remoteEvent))
+                return try! BSONEncoder().encode(documentT)
```
[minor] `try!` -> `try`

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/CoreDocumentSynchronizationConfig.swift:
```diff
@@ -238,15 +264,15 @@ internal struct CoreDocumentSynchronization: Hashable {
+    mutating func setPendingWritesComplete(atVersion: Document) throws {
         docLock.writeLock()
         defer { docLock.unlock() }
         self.lastUncommittedChangeEvent = nil
```
[nit] not your code, but Doug and I spoke with Eric and we decided that we should drive-by rename `lastUncommittedChangeEvent` to `uncommittedChangeEvent` since the "last" is a misnomer. There is only ever one uncommitted change event since they're always coalesced.

dkaminsky replied:
+1 please fix now

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/InstanceSynchronizationConfig.swift:
```diff
@@ -90,15 +98,23 @@ internal struct InstanceSynchronization: Sequence {
+                config.namespaces[namespace] = newConfig.config
+                return newConfig
+            } catch {
+                errorListener?.on(error: error, forDocumentId: nil)
```
[discussion] hmm I see how this can be weird. An error on a nil document ID probably means some fatal error has occurred. 

This would definitely not be in scope for this ticket, but we should maybe think about exposing a fatal error handler to differentiate this from normal sync errors that might freeze a document.

Also to answer your offline question about how we should handle errors that happen as part of CRUD operations. I think sync Android operations just throw, and we wrap that as the exception of the Task. For Swift, we'll probably just wrap anything that throws as the error in the `StitchResult<T>`.

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/MongoNamespace.swift:
```diff
@@ -2,31 +2,22 @@
  * A MongoDB namespace, which includes a database name and collection name.
  */
 public struct MongoNamespace: Codable, CustomStringConvertible, Hashable {
+    private enum CodingKeys: String, CodingKey {
```
[minor] so I'm assuming this works, but then the underlying data structure for synchronization configs will be different on Android and iOS. I think it's important that we somehow maintain the "db.coll" String representation of the namespace.

jsflax replied:
Problem is, that's not how namespaces are represented by our API. I feel that API parity is more important.

adamchel replied:
where do we get a namespace back from our API?

jsflax replied:
ChangeEvents, unless the docs are wrong

adamchel replied:
ah it looks like in Android we decode them manuall in the change event decoder:

https://github.com/mongodb/stitch-android-sdk/blob/master/core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/ChangeEvent.java#L368

adamchel replied:
I think keeping data the same on the device is going to be important as well, especially if we're going to end up going with a common C++ core. It helps us avoid doing local device data migrations in the future

adamchel replied:
So I think we should still encode the MongoNamespace as `"\(db).\(coll)"` in the sync config

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/NamespaceSynchronizationConfig.swift:
```diff
@@ -26,8 +26,9 @@ internal struct NamespaceSynchronization: Sequence {
+
         static func == (lhs: NamespaceSynchronization.Config, rhs: NamespaceSynchronization.Config) -> Bool {
-            return true
+            return lhs.namespace == rhs.namespace
```
[q] when are we actually using the comparison? I think if we don't need it, we should get rid of it instead of having a shallow comparison

jsflax replied:
We may use it down the line, later in our development process. I can remove it for now.

adamchel replied:
yeah let's remove it until we actually need it, and we can decide what makes two configs equal then

dkaminsky replied:
not needed in tests?

adamchel commented on Core/Services/StitchCoreRemoteMongoDBService/Tests/StitchCoreRemoteMongoDBServiceTests/Sync/ChangeEventUnitTests.swift:
```diff
@@ -107,6 +121,61 @@ class ChangeEventUnitTests: XCTestCase {
+                    to: actualLocalDeleteEvent)
+    }
+
+    func testRoundTrip() throws {
```
[minor] can you add code to test the round trip for a change event with an unknown change event? It can be a `"blah"` event going in, and should be an `"unknown"` event coming out.

jsflax replied:
I already did.

adamchel replied:
🤦‍♂️  my bad
